### PR TITLE
Multi-Release-Jar with initial support for JDK 9 and JDK 10 NIO API

### PR DESCRIPTION
**Pull request by mkarg**
Turning Plexus Utils into a Multi-Release-Jar to allow using the *best* API on each platform version available without increasing the minimum platform version.

This PR effectively replaces PRs #112 and #113, as per using NIO API there is no need for these separate PRs (they simply intended to use NIO API).

**Note: Due to the way MRJARs are built, the actual content of the outcoming JAR is limited by the actual JDK release used to build. When using JDK 9, no JDK 10 APIs will be contained, etc..**

### CODE REVIEW

**Comment by michael-o** [comment]
This is a great example I could even use in my project. @rfscholte is the expert here.

**Comment by rfscholte** [comment]
Please also confirm that every accessible method of touched by a test. The most tricky part with MRJAR is to keep all these method signatures in sync. 
And of all methods, how many are adjusted and many still have the same implementation?

**Comment by mkarg** [comment]
> Please also confirm that every accessible method of touched by a test. 

I wonder if somebody knows a *simple* way to let Maven (here: on a JDK 10+ runtime) run tests *multiple* times using several release levels, so it effectively "fakes" being a JDK 8, 9, 10... VM? 

>The most tricky part with MRJAR is to keep all these method signatures in sync.
> And of all methods, how many are adjusted and many still have the same implementation?

I think the best way here would not be to keep them in sync, but instead let the `static` facade call a `private` inner class that *only* contains the variant part, but all common part is in an abstract super class. Sounds complex but is rather easy and safe, and would effectively have us end up with a perfectly clear class structure holding only the *modified* methods. WDYT?

**Comment by rfscholte** [comment]
Right, if the amount of copied code is rather large, it is better to introduce a class just for the specific versions-implementations.

Regarding running the tests: https://maven.apache.org/plugins/maven-compiler-plugin/multirelease.html contains all kinds of strategies, none is perfect. So far this setup + a CI server works best for our projects.

**Comment by mkarg** [comment]
> Right, if the amount of copied code is rather large, it is better to introduce a class just for the specific versions-implementations.

I will refactor the code of this PR in that way, so we have only the actually modified *methods* in the version-specific source folders.

> Regarding running the tests: https://maven.apache.org/plugins/maven-compiler-plugin/multirelease.html contains all kinds of strategies, none is perfect. So far this setup + a CI server works best for our projects.

As the test matrix is already in place, the sole problem we have with this PR is that we need to split the stages for build from the stages for test. In the end, build stage must happen on JDK 10 and higher *only*, and *that* artifact is then to be tested on 8...16. I think that should work. Any better idea?

**Comment by mkarg** [comment]
Refactored code in https://github.com/codehaus-plexus/plexus-utils/pull/120/commits/bc5b0ab5419e8d1a4acd2ed8a4ff158cb325257f. No more complete duplicates of `IOUtil` but only lean overwrites of the actual version-specific methods. WDYT?

**Comment by rfscholte** [comment]
I think we can simplify without using a singleton.
Have one class, e.g. `IOCopyUtil` having the copy-methods, but static.

**Comment by mkarg** [comment]
Refactored in https://github.com/codehaus-plexus/plexus-utils/pull/120/commits/e16dc59a77ef28063084413136407a3f020e5688. I prefer keeping the name `VersionSpecifics` so we can later put more methods from other classes into the same one, so the name should neither be bound to actually the `copy` methods or the `IOUtils` class.

**Comment by mkarg** [comment]
I have modified the Github CI script in https://github.com/codehaus-plexus/plexus-utils/pull/120/commits/cecd035a846940700fbc62c4eb308e3fb2dba19f so it will build with JDK 15 *alway*s but *just* tests with the full matrix, hence the JDK 8 and 9 test is green again now! :-)

If we need to proof that we can *build* (not just *test*) on a full matrix (excluding builds on 8 and 9 as certainly these won't be able to compile Java 10 source code), then I could modify `maven.yml` to contain two separate jobs with *different* matrixes. **Is that wanted / needed?**

**Comment by rfscholte** [comment]
No, this is not good. You can only verify it works with JDK8 by building and testing it with JDK8.
And I see the issue: you're not using JDK activated profiles. See github.com/apache/maven-compiler-plugin/blob/master/src/it/multirelease-patterns/singleproject-runtime/pom.xml to see how it should work.

**Comment by mkarg** [comment]
Why can't I see if the JAR works on JDK 8 when I build it on 15 but test it with 8? For what is the test good then...?

That makes no sense to me since the result of that would be that on JDK 9 you end up with a JAR that would contain the variants for 8 and 9 but not for 10. So if we do not build with 10+ always, you would effectively get *different* content in the JAR. Why should somebody want that?

**Comment by rfscholte** [comment]
How do you test it with JDK8?

**Comment by mkarg** [comment]
I don't understand the question. As you can see in https://github.com/codehaus-plexus/plexus-utils/pull/120/checks?check_run_id=1640215941#step:7:12, the current CI matrix tests on Java 8, but it uses the JAR built using JDK 15 for the tests (see https://github.com/codehaus-plexus/plexus-utils/pull/120/checks?check_run_id=1640215941#step:5:10). This works pretty well as long as what you want to test is whether a JDK 15 built JAR *works (i. e. finds the default implementation)* on JDK 8.

**Comment by mkarg** [comment]
@rfscholte WDYT?

**Comment by rfscholte** [comment]
please just use JDK activated profiles. This will also make it easier to test locally.

**Comment by mkarg** [comment]
@rfscholte According to your wish, I have undone the Gitlab CI change, so now you can locally build using JDK 8 or 9 to get *different* multi-release JARs for those platforms (while still the general idea is to build on JDK 10+ to get the sole one to be published on Maven Central). Is it now in the way that you prefer?

**Comment by mkarg** [comment]
Everybody please vote! I want to merge this in the next days! Thanks! :-)

File: src/main/java10/org/codehaus/plexus/util/VersionSpecifics.java
Comment (rfscholte) [inline src/main/java10/org/codehaus/plexus/util/VersionSpecifics.java:15]: I'd say util, not singleton

File: src/main/java/org/codehaus/plexus/util/VersionSpecifics.java
Comment (rfscholte) [inline src/main/java/org/codehaus/plexus/util/VersionSpecifics.java:17]: I'd say util, not singleton

File: src/main/java/org/codehaus/plexus/util/VersionSpecifics.java
Comment (rfscholte) [inline src/main/java/org/codehaus/plexus/util/VersionSpecifics.java:12]: I still don't like the classname. It should contain something with IO, because other `o.c.p.util` classes might need a version specific implementation too, that shouldn't be mixed with this one.

**Review by rfscholte — CHANGES_REQUESTED** [review]

File: src/main/java10/org/codehaus/plexus/util/VersionSpecifics.java
Comment (mkarg) [reply src/main/java10/org/codehaus/plexus/util/VersionSpecifics.java:15]: >I'd say util

I actually don't understand what you want to tell me with "I'd say util".

File: src/main/java/org/codehaus/plexus/util/VersionSpecifics.java
Comment (mkarg) [reply src/main/java/org/codehaus/plexus/util/VersionSpecifics.java:12]: In fact I do not understand why you do not want to mix it. Having all tweaks in a single class makes it rather easy to find all the version specifics very easily.

Comment (rfscholte) [reply src/main/java/org/codehaus/plexus/util/VersionSpecifics.java:12]: Suppose we can have improvements by using TextBlock for some StringUtils. We would need to introduce a new VersionSpecifics, and it will contain clones of current copy methods, while existing VersionSpecifics need to be extended with the newly introduces String methods.
Maintaining MRJAR code is already hard, this will even make it harder.

Comment (mkarg) [reply src/main/java/org/codehaus/plexus/util/VersionSpecifics.java:12]: There is actually no need for "a new VersionSpecifics". The idea behind my code actually is that Plexus-Utils JUST has that single `VersionSpecifics` class for *all* packages, so maintaining the MRJARs will become a single-location task, hence is *pretty easy* to do: Just put a new `static` **method** in the existing `VersionSpecifics`, and make the original code call it. I cannot see what is complex with that, and in fact, what you counter-propose makes it much, much more complex, as you will have *lots* of version specific classes in the end, *and* nobody would find them as it is totally unclear why the implementation of version specifics should be named "util", as these are not utilities *at all* but actually are what they are: version specifics. Sorry, I cannot see your point here at all.

**Comment by mkarg** [comment]
@rfscholte Robert, to make progress I'd like to kindly ask you to answer to my above comments. Thanks.

**Comment by rfscholte** [comment]
There's nothing more to explain. I did my best, but you don't seem to understand. 
On clue that you might be missing is that at runtime only one class will be loaded, the best matching versioned class. If this class is missing a method, it'll give a runtime exception. It won't try to load the second best class to see if the method exists there.

**Comment by mkarg** [comment]
> There's nothing more to explain. I did my best, but you don't seem to understand.

I do respect your opinion and merits, but honestly, you did *not answer my question* what you mean by your phrase "I'd say util", and as you *request* a change here, whom else shall I ask for what *you* wanted to say with "I'd say util"? Does that mean, I *must* do something here or it is just a proposal, and what shall the proposal be, to *rename* that class or change a comment, or...?

> On clue that you might be missing is that at runtime only one class will be loaded, the best matching versioned class. If this class is missing a method, it'll give a runtime exception. It won't try to load the second best class to see if the method exists there.

This is a side effect of the way *you* proposed to build the MRJAR. So fix it, we must change the POM to always compile the full source code, which is a deviation of *your* proposal. If you *insist* on that, please clearly tell, *or* clearly tell what you want *otherwise*. Thanks.

**Comment by mkarg** [comment]
@michael-o Maybe it makes sense that you explain to me in German in short Zoom chat what Robert wants. I just don't understand him. Would that be ok or do you have another idea how to go on from here? I want to fix all that is *needed*, but for that, I need to *understand* what Robert wants.

**Comment by rfscholte** [comment]
I've been thinking how to solve this as concrete as possible.
What I would like to see is the introduction of `BaseIOUtil`, having `IOUtil extends BaseIOUtil`.
So roughly renaming `VersionSpecifics` to `BaseIOUtil`, and changing the calls to `VersionSpecifics.x()` to `super.x()` would fix this.

**Comment by mkarg** [comment]
@rfscholte I changed it in the way you requested. So do you give this code a +1 now? :-)

**Comment by rfscholte** [comment]
this is fine

**Review by olamy — CHANGES_REQUESTED** [review]
As far I understood it's now required to build/release the project with at least java10 otherwise some features will not be build/deploy. But I cannot see anything enforcing this neither documenting it.

**Comment by olamy** [comment]
@mkarg furthermore can you add more details in the pr title on what this pr is doing because it's more than "Multi-Release-Jar" :)

**Comment by mkarg** [comment]
> As far I understood it's now required to build/release the project with at least java10 otherwise some features will not be build/deploy. But I cannot see anything enforcing this neither documenting it.

I will look into that.

**Comment by mkarg** [comment]
> @mkarg furthermore can you add more details in the pr title on what this pr is doing because it's more than "Multi-Release-Jar" :)

Uhm... actually it is *only* doing that. Don't know what you're talking about...?

**Comment by olamy** [comment]
> > @mkarg furthermore can you add more details in the pr title on what this pr is doing because it's more than "Multi-Release-Jar" :)
> 
> Uhm... actually it is _only_ doing that. Don't know what you're talking about...?

Sorry I have been confusing by the description mentioning 
> This PR effectively replaces PRs #112 and #113 if it is accepted by a majority.

So in my mind it was some new features using NIO9 and NIO10

**Comment by mkarg** [comment]
IIUC then all change requests are resolved so I would merge in the next step. Anybody -1 for that plan?

**Comment by mkarg** [comment]
As nobody responed within five business days, I assume everybody is fine with my proposed merge, hence I will do that now.

**Comment by olamy** [comment]
What??? what has been done regarding https://github.com/codehaus-plexus/plexus-utils/pull/120#issuecomment-767695517 ??? If some jdk level is required we must enforce it via some enforcer rules to provide easy human readable message.

**Comment by olamy** [comment]
And seriously such change must be squashed into a single when merged in master branch.

**Comment by olamy** [comment]
I'm definitely -1 on such changes!!!

**Comment by olamy** [comment]
can you please fix that!!

**Comment by michael-o** [comment]
Merging dirty commits is a no go.

**Comment by olamy** [comment]
@michael-o and it's too late now...

**Comment by michael-o** [comment]
> @michael-o and it's too late now...

True, I don't understand that

**Comment by mkarg** [comment]
Sorry if I did something wrong, but **nobody** was responding since **five** days, even after explicitly asking for comments, so I thought you are all fine (just as Robert was, who explicitly answered earlier). I do not see that an enforcer is needed, as no particular JDK level is actually *needed*. The solution works well with *any* JDK level, and the description of the PR clearly explains how it works. No harm is done actually. Whether or not you squash changes is a question of likes and dislikes. Neither did somebody request that, not did I find a written rule about that, and in many projects "dirty" commits for accepted in SNAPSHOTs. So if there is such a rule, and if you get asked by a new committer about comments, then you maybe should refer to it and request a change *when asked for*. I did not merge "hastily", I asked *days ago*. Sorry guys, but staying silent for five days was simply misunderstood.

**Comment by mkarg** [comment]
Guys, I am taking your comment very seriously so I checked the points you mentioned today. In fact, to me comment https://github.com/codehaus-plexus/plexus-utils/pull/120#issuecomment-767695517 *was* resolved by me changing the PR description and nobody responded to that, so I thought you are fine. The comments, besides two, are *independent*, and not dirty. For me, supporting Java 9, 10, MJARs, and changing things later due to Robert's request, are *neither* intermediate nor dirty commits, so actually I thought it is *good* to have them separate (some OS projects even prevent squashing but insist on having the full history of each merge branch). "dirty" is only the support for JDK 15, and yes, *if* you would have told me *earlier* (you had time since Janary 16 and was asked explicitly five days ago) I certainly would have followed your request.

So how to go on from here? Do you want me change change something *now* then please exactly tell me what. If there is a document that I shall read to prevent future "surprises" then please directly point me to that link.

Thanks guys, and sorry again.

**Comment by michael-o** [comment]
The problem ist that this topic is complex, I am not qualified to judge here. Most aren't. Five days sounds a lot, but it isn't. Most need weeks to find time for this. I am slow.

**Comment by mkarg** [comment]
I totally understand that, but the truth is, the PR is open since January 1st and the discussions werde finished in January 26th, so this is much more than five days. The five days were just a final reminder. Robert is the defintive expert, and he was fine with the solution provides by January 26th already.

**Comment by michael-o** [comment]
If @rfscholte was fine with that, I am too. But a cleanup is imperative.

**Comment by olamy** [comment]
> Sorry if I did something wrong, but **nobody** was responding since **five** days, even after explicitly asking for comments, so I thought you are all fine (just as Robert was, who explicitly answered earlier). I do not see that an enforcer is needed, as no particular JDK level is actually _needed_. The solution works well with _any_ JDK level, and the description of the PR clearly explains how it works. No harm is done actually. Whether or not you squash changes is a question of likes and dislikes. Neither did somebody request that, not did I find a written rule about that, and in many projects "dirty" commits for accepted in SNAPSHOTs. So if there is such a rule, and if you get asked by a new committer about comments, then you maybe should refer to it and request a change _when asked for_. I did not merge "hastily", I asked _days ago_. Sorry guys, but staying silent for five days was simply misunderstood.


Sorry but you are wrong! Enforcer rule is definitely needed as asked here: https://github.com/codehaus-plexus/plexus-utils/pull/120#pullrequestreview-574998126

If I build the project with java11 or java8 or java9 the produced jar will have different content.
with java8, everything related to java9 or java10 profile are not included.
As we do not enforce the java version, we are not sure about what will be released it will depends on the jdk used by the developer releasing the project.
Sorry but it's totally wrong!

jdk 11 build jar content: 
```
unzip -l target/plexus-utils-3.4.0-SNAPSHOT.jar | grep version
        0  01-20-2020 18:52   META-INF/versions/
        0  01-20-2020 18:52   META-INF/versions/10/
        0  01-20-2020 18:52   META-INF/versions/10/org/
        0  01-20-2020 18:52   META-INF/versions/10/org/codehaus/
        0  01-20-2020 18:52   META-INF/versions/10/org/codehaus/plexus/
        0  01-20-2020 18:52   META-INF/versions/10/org/codehaus/plexus/util/
        0  01-20-2020 18:52   META-INF/versions/9/
        0  01-20-2020 18:52   META-INF/versions/9/org/
        0  01-20-2020 18:52   META-INF/versions/9/org/codehaus/
        0  01-20-2020 18:52   META-INF/versions/9/org/codehaus/plexus/
        0  01-20-2020 18:52   META-INF/versions/9/org/codehaus/plexus/util/
      843  01-20-2020 18:52   META-INF/versions/10/org/codehaus/plexus/util/BaseIOUtil.class
      942  01-20-2020 18:52   META-INF/versions/9/org/codehaus/plexus/util/BaseIOUtil.class
```

jdk8 build jar content:
```
unzip -l target/plexus-utils-3.4.0-SNAPSHOT.jar | grep version
mbp-olamy-2:plexus-utils olamy$ 

```

Your comment on no rule written on squash etc... Well it's really a common sense as this pr just have different commits with non related comments... 

5 days. what is 5 days in a world with global pandemic, people working on different timezone and with easter weekend (school holidays which can change a lot the availability of some contributors)

**Comment by mkarg** [comment]
@olamy Sorry but you are wrong in some points, and I would kindly ask you to check my original commit for MJAR support instead of just looking at the final PR.

- It was not just *five* days, it was months. See the actual discussion above and my explanation from yesterday https://github.com/codehaus-plexus/plexus-utils/pull/120#issuecomment-812898424.
- Originally I proposed using JDK 15 to be *mandatory* (that's why I originally *added* it here https://github.com/codehaus-plexus/plexus-utils/pull/120/commits/cecd035a846940700fbc62c4eb308e3fb2dba19f and my intention *was* to enforce it).
- It was Robert who explicitly asked to *not* enforce JDK 15, but to in fact make the build work with JDK 8. That's why I have undone that commit. This is what you call "dirty" but it is needed to *understand* the history and to *prevent* the current discussion.
- The PR *must* keep most of its commits standalone, as support for NIO 9 and NIO 10 won't work without support for MJARs, but actually are standalone *ideas* (that's why there originally were separate PRs for both), and support for MJARs make no sense without support for NIO 9 at least. The sole thing I *could* squash would be the history that proofs that I originally *had* enforced JDK 15, but then Robert wanted support for JDK 8. It seems, you didn't follow the PR's discussion, but in fact, it clearly explains why the PR is in the current state and why I merged it, and that it was *not my* idea to keep JDK 8, and that I explicitly warned Robert about the fact that in the end the changes *he* wanted lead to the fact that depending und the instaled JDK the result will be different. This all was *his* idea, not *mine*. Please read the full thread!

I deliberately did not squash it and deliberately waited for *months* for you to be able to follow exactly that situation.

So until you find an agreement with Robert, I simply *cannot* fix this, because what *he* wanted me to do is exactly the opposite of what *you* want me to do, and *he* was very eager about this discussion, while you had time since January 26th to respond but just didn't.

**Comment by rfscholte** [comment]
To ensure the the Java 8 classes still work with Java 8, you must run it with JDK8. That's why the CI server is important, you cannot verify this with a single Maven build on your local system.
However, for releasing you MUST use the highest required JDK to build the code. To do that you must add this as an enforcer rule to the release profile.

**Comment by mkarg** [comment]
So where to go from here, just add a commit ontop which adds the enforcer rule on the release profile, *or* squashing *everything* (i. e. breaking the branch's history) *or* just squashing the changes you personally requested (which still breaks history)? I need clear instructions how it is definitively to be done, since (as I said) this is *not* common sense and done differently in other open source projects. Thanks.

**Comment by olamy** [comment]
> @olamy Sorry but you are wrong in some points, and I would kindly ask you to check my original commit for MJAR support instead of just looking at the final PR.
> 
> * It was not just _five_ days, it was months. See the actual discussion above and my explanation from yesterday [#120 (comment)](https://github.com/codehaus-plexus/plexus-utils/pull/120#issuecomment-812898424).
> * Originally I proposed using JDK 15 to be _mandatory_ (that's why I originally _added_ it here [cecd035](https://github.com/codehaus-plexus/plexus-utils/commit/cecd035a846940700fbc62c4eb308e3fb2dba19f) and my intention _was_ to enforce it).

java15 as a minimun? really? not sure the IT industry has reached such level.

> * It was Robert who explicitly asked to _not_ enforce JDK 15, but to in fact make the build work with JDK 8. That's why I have undone that commit. This is what you call "dirty" but it is needed to _understand_ the history and to _prevent_ the current discussion.

who cares about having this in a code history. We are not here to store some fingers pointing. It's already part of the PR discussion no really need to add more.

> * The PR _must_ keep most of its commits standalone, as support for NIO 9 and NIO 10 won't work without support for MJARs, but actually are standalone _ideas_ (that's why there originally were separate PRs for both), and support for MJARs make no sense without support for NIO 9 at least. The sole thing I _could_ squash would be the history that proofs that I originally _had_ enforced JDK 15, but then Robert wanted support for JDK 8. It seems, you didn't follow the PR's discussion, but in fact, it clearly explains why the PR is in the current state and why I merged it, and that it was _not my_ idea to keep JDK 8, and that I explicitly warned Robert about the fact that in the end the changes _he_ wanted lead to the fact that depending und the instaled JDK the result will be different. This all was _his_ idea, not _mine_. Please read the full thread!

As I read this comment, some commits cannot work separately so better to have a single one.
More generally it's just better to have a new feature including in a single commit to be able to easily revert it if needed.


> 
> I deliberately did not squash it and deliberately waited for _months_ for you to be able to follow exactly that situation.
> 
> So until you find an agreement with Robert, I simply _cannot_ fix this, because what _he_ wanted me to do is exactly the opposite of what _you_ want me to do, and _he_ was very eager about this discussion, while you had time since January 26th to respond but just didn't.

Perso I don't see adding this build complexity (debugging) as a real step forward. Does it really worth? Is it really so much performant?
Anyway java9 and java10 are dead release. I would prefer to have only java11.

**Comment by mkarg** [comment]
@olamy JDK 15 was initially proposed as the enforced build JDK as it was the latest GA release available on Github, so it would allow to add support for more version folders in the MJAR without changing the POM; anyways I just told you that to proof that *I did have* the idea of enforcing the JDK version used for the build (just as you do request it). Hence I have no problem with reducing the enforcement to just JDK 11 at all, I'm just waiting for instructions as you see in the PR comments. Whether something is a dead horse plays no role for MJARs: These are the *minimum* levels the API changes will work upon, so in fact, we do currently support 8, 9 and 10, but actually 8, 9, 10, 11, 12, 13, 14, 15, 16, 17-EA with this very same solution. "Just 11" would not work as @rfscholte insisted on having support for 8, too (see the PR history). And yes, it is worth the complexity, and we should concentrate our discussion on going forward not backwards.

@rfscholte As I do not want to do a second failure in your unwritten processes, **I need clear instructions whether I may directly rewrite the history** (as apparently targeted by @olamy IIUC his complaint) or **whether I may push the needed commit for the enforcer just ontop**, or **whether I shall provide a new PR** with just that single commit. Or whatever else this community wants me to do to stop the discussion about the past and go in the right direction for the future.

See, now that I learned that nobody will point a finger on anybody (which I really appreciate) I really want to help fixing the "mess" as best as I can, but **I need to know *which* way the majority wants me to go*. Thanks.

**Comment by rfscholte** [comment]
This is the first time I hit this situation, so I'd follow Oliviers advice here.

**Comment by olamy** [comment]
please provide a PR.
can you please provide some jmh tests to prove it really worths such complexity?

**Comment by mkarg** [comment]
> please provide a PR. rewriting master branch history is just worst.
> can you please provide some jmh tests to prove it really worths such complexity?

I will provide a PR for the enforcer.

Regarding JMH I will look into something on my Youtube channel regarding performance in general.

**Comment by olamy** [comment]
> > please provide a PR. rewriting master branch history is just worst.
> > can you please provide some jmh tests to prove it really worths such complexity?
> 
> I will provide a PR for the enforcer.
> 
> Regarding JMH I will look into something on my Youtube channel regarding performance in general.

take it easy buddy we do not need some TV show but just few jmh tests....

**Comment by mkarg** [comment]
> > Regarding JMH I will look into something on my Youtube channel regarding performance in general.
> take it easy buddy we do not need some TV show but just few jmh tests....

comparing old-school code with NIO 10 code sounds like a good topic for a shootout video, and I am always looking for interesting topics for my channel anyways

**Comment by mkarg** [comment]
> can you please provide some jmh tests to prove it really worths such complexity?

@olamy It took some time as I was rather busy, but meanwhile I found the time to provide the JMH benchmark.

Please find https://gitlab.com/mkarg/jaxrs-done-right/-/blob/1a85b315470af7bd5702dd6762d912c0f053f5ca/src/main/java/eu/sandbox/Sandbox.java.

Ignoring all the out-commented benchmarks, it effectively just compares what Plexus-Utils did before (using NIO 7) and after (using NIO 9 and NIO 10) my merge of the MRJAR PR. 

I have performed this benchmark on both, Linux and Windows, and while certainly the difference currently is not dramatic yet, I personally would say the complexity (which I, in fact, would *not* say is really big) *is* worth the improved throughput, in particular as the rather complex implementation of the transferTo() methods clearly could be further optimized, while the explicit custom code that one had to write in lieu of transferTo() simply already reached the peak optimization that is possible outside of the JRE.

The actual numbers of my Windows Laptop are:

```
Sandbox.java07_Files_newBufferedReader        1000  thrpt   25  3257,480 ▒  75,403  ops/s
Sandbox.java07_Files_newBufferedReader     1000000  thrpt   25   518,090 ▒  14,041  ops/s
Sandbox.java07_Files_newBufferedReader  1000000000  thrpt   25     0,472 ▒   0,008  ops/s
Sandbox.java07_Files_newInputStream           1000  thrpt   25  3446,345 ▒ 197,931  ops/s
Sandbox.java07_Files_newInputStream        1000000  thrpt   25  1136,500 ▒  24,773  ops/s
Sandbox.java07_Files_newInputStream     1000000000  thrpt   25     1,137 ▒   0,023  ops/s
Sandbox.java09_InputStream_transferTo         1000  thrpt   25  3332,082 ▒  72,743  ops/s
Sandbox.java09_InputStream_transferTo      1000000  thrpt   25   865,717 ▒  24,389  ops/s
Sandbox.java09_InputStream_transferTo   1000000000  thrpt   25     0,733 ▒   0,003  ops/s
Sandbox.java10_Reader_transferTo              1000  thrpt   25  3141,430 ▒  71,134  ops/s
Sandbox.java10_Reader_transferTo           1000000  thrpt   25   545,857 ▒  15,744  ops/s
Sandbox.java10_Reader_transferTo        1000000000  thrpt   25     0,489 ▒   0,004  ops/s
```

Some actual numbers of a Linux cloud container:

```
Sandbox.java07_Files_newBufferedReader        1000  thrpt    5  31405.640 ±  3084.986  ops/s
Sandbox.java07_Files_newBufferedReader     1000000  thrpt    5    475.655 ±   107.712  ops/s
Sandbox.java07_Files_newBufferedReader  1000000000  thrpt    5      0.432 ±     0.017  ops/s
Sandbox.java07_Files_newInputStream           1000  thrpt    5  63765.430 ± 15673.334  ops/s
Sandbox.java07_Files_newInputStream        1000000  thrpt    5   1549.640 ±   150.097  ops/s
Sandbox.java07_Files_newInputStream     1000000000  thrpt    5      1.087 ±     0.110  ops/s
Sandbox.java09_InputStream_transferTo         1000  thrpt    5  75235.225 ±  9918.075  ops/s
Sandbox.java09_InputStream_transferTo      1000000  thrpt    5   1344.171 ±    91.492  ops/s
Sandbox.java09_InputStream_transferTo   1000000000  thrpt    5      0.996 ±     0.031  ops/s
Sandbox.java10_Reader_transferTo              1000  thrpt    5  36727.881 ±   684.608  ops/s
Sandbox.java10_Reader_transferTo           1000000  thrpt    5    492.899 ±    24.666  ops/s
Sandbox.java10_Reader_transferTo        1000000000  thrpt    5      0.442 ±     0.026  ops/s
```

What these numbers do tell us is that the differences between NIO 7, 9 and 10 are *different* on Windows and Linux (besides the fact that one *cannot* compare a Windows Laptop to a Linux cloud container, certainly), but also, and this is much more important, `Reader.transferTo(Writer)` of NIO 10 *outperforms* custom array-copy-loops of NIO 7 (while being more comfortable). And yes, I also panicked when I noticed that `InputStream.transfertTo(OutputStream)` of NIO 9 actually is *much* slower than custom array-copy-loops of NIO 7! As it happens always, and as it happens on Windows Laptop *and* Linux cloud container, my next step is debugging into the OpenJDK source code what the heck is going on there and possibly proposing a bug fix *there*.

If that means that you want to undo NIO 9 support until then, feel free... but I personaly think we should not workaround specific performance bugs of particular JDK implementations but instead fix them, as we do not know if they even exist in JDK 17 or on ARM or on Mac...

**Comment by mkarg** [comment]
@olamy More  performance news. Looked into the JRE source and did some more benchmarks. Actually the performance drop of `InputStream.transferTo` reported above is not bug at all. It is simply the reflection of the fact that the original Plexus code used a slightly bigger buffer when copying files, effectively making the Plexus code being optimized for larger files than the optimization chosen inside of the JRE itself which is for good for smaller files. It is doubtful what the better optimization goal is. Anyways, I will discuss with the OpenJDK team how we can make `InputStream.transferTo` perform better, or even use more offloading (I did a benchmark with an offloading prototype effectively outperforming *any other* algorithm, including the original Plexus code, unambigously. So I would say, supporting `InputStream.transferTo` might be a step back for *some* file sizes *temporarily*, but has the potential to outperform in future, and I am pretty sure the OpenJDK team will adopt a feasible optimization. Stay tuned.

**Comment by mkarg** [comment]
@olamy Performance update: Filed [enhancement proposal](https://bugs.openjdk.java.net/browse/JDK-8265891) meanwhile, now waiting for OpenJDK's answer. If they accept it, `InputStream.transferTo` will be the fastest possible solution at all, as the proposal uses NIO channels and buffers which turned out to *actually* completely offload to the OS, and outperformed all other methods in my JMH benchmark. Exciting! :-)

**Pull request merged**
